api/eds.proto

```diff
@@ -0,0 +1,178 @@
+  // 4. For each cluster, Envoy load balances incoming traffic to upstream hosts
+  //    based on per-zone weights and/or per-instance weights (if specified)
+  //    based on intra-zone LbPolicy.
+  // 5. When upstream hosts reply, they optionally add header <define header
+  //    name> with ASCII representation of EndpointLoadMetricStats.
+  // 6. Envoy aggregates load reports over the period of time given to it in
+  //    LoadBalanceResponse.load_reporting_interval. This includes aggregation
+  //    stats Envoy maintains by itself (total_requests, rpc_errors etc.) as
+  //    well as load metrics from upstream hosts.
+  // 7. When the timer of load_reporting_interval expires, Envoy sends new
+  //    LoadBalanceRequest filled with load reports for each cluster.
+  // 8. The management server uses the load reports from all reported Envoys
+  //    from around the world, computes global assignment and prepares traffic
+  //    assignment destined for each zone Envoys are located in. Goto 3.
+  // TODO(htuch): Add @amb67's diagram.
+  rpc BalanceLoad(stream LoadBalanceRequest)
+      returns (stream LoadBalanceResponse) {
+  }
+}
+
+message EndpointDiscoveryRequest {
+  repeated string cluster_name = 1;
+  Node node = 2;
+}
+
+message LbEndpoint {
+  EndpointHealth endpoint_health = 1;
+  // The optional canary status of the upstream host. Envoy uses the canary
+  // status for various statistics and load balancing tasks documented
+  // elsewhere.
+  google.protobuf.BoolValue canary = 2;
+  // The optional load balancing weight of the upstream host, in the range 1 -
+  // 100. Envoy uses the load balancing weight in some of the built in load
+  // balancers. All load balancing weights in a locality must sum to 100%.
+  // The effective load balancing weight is load_balancing_weight multiplied by
+  // the locality wide load_balancing_weight. If unspecified, each host is
+  // presumed to have equal weight in a locality.
+  google.protobuf.UInt32Value load_balancing_weight = 3;
+}
+
+// All endpoints belonging to a Locality.
+message LocalityEndpoints {
+  Locality locality = 1;
+  repeated LbEndpoint lb_endpoints = 2;
+
+  // Optional: Per region/zone/sub_zone weight - range 1-100. All load balancing
+  // weights in a cluster must sum to 100%. If unspecified, each locality is
+  // presumed to have equal weight in a cluster.
+  google.protobuf.UInt32Value load_balancing_weight = 3;
+}
+
+// Each route from RDS will map to a single cluster (or specify its own weights
+// for WeightedCluster which are discoverable by RDS). In EDS, each cluster is
+// treated independently from a LB perspective, with LB taking place between the
+// Localities within a cluster and at a finer granularity between the hosts
+// within a locality.
+message ClusterEndpoints {
+  string cluster_name = 1;
+  repeated LocalityEndpoints locality_endpoints = 2;
+}
+
+message EndpointDiscoveryResponse {
+  repeated ClusterEndpoints cluster_endpoints = 1;
```

Comment:
why doesn't the REST version return ClusterLoadBalance like the streaming version?